enhancements/machine-api/aws-efa-networking.md

```diff
@@ -0,0 +1,357 @@
+---
+title: aws-efa-networking
+authors:
+  - "@JoelSpeed"
+reviewers:
+  - "@elmiko"
+  - "@deads2k"
+  - "@sdodson"
+  - "@danwinship"
+approvers:
+  - "@deads2k"
+  - "@sdodson"
+  - "@knobunc"
+api-approvers:
+  - "@deads2k"
+creation-date: 2021-12-15
+last-updated: 2021-12-15
+tracking-link:
+  - https://issues.redhat.com/browse/OCPCLOUD-1353
+---
+
+# AWS EFA Networking
+
+## Summary
+
+AWS Elastic Fabric Adapter (EFA) is a feature within EC2 that allows for network adapters that provide high throughput
+for HPC applications communicating via MPI to be attached to certain instance types. With the presence of these network
+adapters and some additional drivers, workloads can take advantage of greatly improved network performance between
+nodes within the same AWS availability zone.
+
+## Motivation
+
+We have customers already using this feature of EC2 within OpenShift UPI clusters. They have taken on the responsibility
+of making sure the prerequisites are fulfilled and have manually created instances with the attached adapters.
+However, there is a desire to enable this feature to be used in conjunction with existing autoscaling features within
+OpenShift. To enable this, we must provide support for the feature within the Machine API.
+
+### Goals
+
+- Enable Non Control Plane Machines to attach EFA network interfaces as their primary network interface
+- Enable autoscaling of Machines that use EFA network interfaces
```

Comment:
I think this is currently unknown, I imagine it will need a special driver similar to how GPUs work as it is an allocatable resource in the same regard

Reply:
i think we should update this goal if we are not planning to add autoscaling initially.

Reply:
Will the Autoscaler require a special topology aware info to allocate the new node in proximity to existing ones? there's no point on scaling X number of nodes with EFA "enabled" if they are scattered around zones. how can this be ensured when scaling up?

Reply:
The cluster autoscaler is topology aware already, so I would expect the workloads to request to be placed in a single AZ using node affinity, or to be placed close to other similar pods using pod affinity, either should allow the workloads to be placed close together and have the autoscaler understand the requirement

Reply:
i don't think this will be a problem for our implementation of the autoscaler. since we control the node groups based on MachineSets, and those machinesets will carry all the placement info. so long as the user has defined their workloads to use selectors for the machinesets that utilize the placement and efa, it should work as expected.